js/binance.js

```diff
@@ -1357,17 +1385,51 @@ module.exports = class binance extends Exchange {
         return result;
     }
 
+    market (symbol) {
+        // helper method for legacy symbols
+        // this means that 'BTC/USDT' will continue to work for futures
+        if (symbol in this.markets) {
+            // 99% of calls should take this path and would not be slowed down
+            return this.markets[symbol];
+        } else {
+            const defaultType = this.getDefaultType ();
+            if (defaultType === 'linear') {
+                if (symbol.indexOf (':') === -1) {
+                    const parts = symbol.split ('/');
+                    const quote = parts[1];
+                    symbol = symbol + ':' + quote;
+                }
+            } else if (defaultType === 'inverse') {
+                if (symbol.indexOf (':') === -1) {
+                    const parts = symbol.split ('/');
+                    const base = parts[1];
+                    symbol = symbol + ':' + base;
+                }
+            }
+            return super.market (symbol);
+        }
+    }
+
     async fetchMarkets (params = {}) {
-        const defaultType = this.safeString2 (this.options, 'fetchMarkets', 'defaultType', 'spot');
-        const type = this.safeString (params, 'type', defaultType);
+        const types = this.safeValue (this.options, 'fetchMarkets', [ 'spot', 'linear', 'inverse' ]);
+        let result = [];
+        for (let i = 0; i < types.length; i++) {
+            const type = types[i];
+            const response = await this.fetchMarketsByType (type, params);
+            result = this.arrayConcat (result, response);
+        }
+        return result;
+    }
+
+    async fetchMarketsByType (type, params = {}) {
         const query = this.omit (params, 'type');
```

Comment:
This type-fetching logic does not seem right to me. 
First, we read the type from `this.options.fetchMarkets` and return a default array. However, we're ignoring the type in params or the type set in `defaultType` completely. Regarding the first one, it's only being omitted from the query, and the type in `defaultType` is never read, so currently `params/defaultType` have no effect at all at fetching markets

Now, let's assume that fetchMarketsByType uses the type from `params/defaultType` properly and the user did not set `this.options.fetchMarkets` so we have the default value there (`[ 'spot', 'linear', 'inverse' ]`)

we would be doing something like this:

```Javascript
for (i=0; i< [spot,linear,inverse].length) {
       fetchMarketsByType (type, params = {}) {
                actualType = this.safeString(this.options, 'defaultType); // actualType = 'spot'
                 (...)
       }
}
```

As you can see we're potentially calling the same thing three times instead of one, so I would suggest moving all of this type guessing logic (check params, check options, check default, get the actual type/types to fetch) to the method above (fetchMarkets) and `fetchMarketsByType` would only respect the  single type received as argument